### PR TITLE
Add step-by-step and pipeline tutorials for reinforcement learning with Vertex AI.

### PR DESCRIPTION
**Pull request by KathyFeiyang**
### Add step-by-step and pipeline tutorials for reinforcement learning with Vertex AI.

Add two reinforcement learning on Vertex AI prototypes. The prototypes use [TF-Agents](https://www.tensorflow.org/agents), [Kubeflow Pipelines (KFP)](https://www.kubeflow.org/docs/components/pipelines/overview/pipelines-overview/) and [Vertex AI](https://cloud.google.com/vertex-ai) in building a reinforcement learning application: movie recommendation system based on the MovieLens 100K dataset.

- **Step-by-step demo**: showcase how to use custom training, custom hyperparameter tuning, custom prediction and endpoint deployment of Vertex AI to build a RL movie recommendation system

- **End-to-end pipeline demo**: showcase how to build a RL-specific MLOps pipeline using KFP and [Vertex Pipelines](https://cloud.google.com/vertex-ai/docs/pipelines), as well as additional Vertex AI and GCP services such as [BigQuery](https://cloud.google.com/bigquery), [Cloud Functions](https://cloud.google.com/functions), [Cloud Scheduler](https://cloud.google.com/scheduler), [Pub/Sub](https://cloud.google.com/pubsub).

Each demo contains a notebook that carries out the full workflow and user instructions, and a `src/` directory for Python modules and unit tests.

---
Before submitting a Jupyter notebook, follow this mandatory checklist:

- [x] Use the [notebook template](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/master/notebooks/notebook_template.ipynb) as a starting point.
- [x] Follow the style and grammar rules outlined in the above notebook template.
- [x] Verify the notebook runs successfully in Colab since the automated tests cannot guarantee this even when it passes.
- [x] [N/A] Passes all the required automated checks
- [x] [N/A] You have consulted with a tech writer to see if tech writer review is necessary. If so, the notebook has been reviewed by a tech writer, and they have approved it.
- [x] This notebook has been added to the CODEOWNERS file, pointing to the author or the author's team. If the CODEOWNERS file doesn't exist, create one in the nearest folder that makes sense.
- [x] The Jupyter notebook cleans up any artifacts it has created (datasets, ML models, endpoints, etc) so as not to eat up unnecessary resources.

### CODE REVIEW

**Comment by yinghsienwu** [comment]
LGTM! Thanks for the PR.

File: tutorials/community/reinforcement_learning/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Reinforcement Learning Protytopes on Vertex AI
```
Comment (ivanmkc) [inline tutorials/community/reinforcement_learning/README.md:1]: Spelling of "prototype" is wrong

**Review by ivanmkc — APPROVED** [review]
LGTM because community tutorials don't need review.

Just add the CODEOWNERS file

File: tutorials/community/reinforcement_learning/pipeline_reinforcement_learning_vertex_ai/src/trainer/trainer_component.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+def training_op(
+    training_artifacts_dir: str,
+    tfrecord_file: str,
+    num_epochs: int,
+    rank_k: int,
+    num_actions: int,
+    tikhonov_weight: float,
+    agent_alpha: float
+) -> NamedTuple("Outputs", [
+    ("training_artifacts_dir", components.OutputPath),
```
Comment (Ark-kun) [inline tutorials/community/reinforcement_learning/pipeline_reinforcement_learning_vertex_ai/src/trainer/trainer_component.py:31]: This `components.OutputPath` syntax is invalid.

Just replace it with `str`

File: tutorials/community/reinforcement_learning/pipeline_reinforcement_learning_vertex_ai/pipeline_reinforcement_learning_vertex_ai.ipynb [hunk 3]
```diff
@@ -0,0 +1,1780 @@
...
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "0YwZd9-4AyCK"
+      },
+      "outputs": [],
+      "source": [
+        "! gsutil cp gs://cloud-samples-data/vertex-ai/growth-rl-demo/u.data $RAW_DATA_PATH"
```
Comment (Ark-kun) [inline tutorials/community/reinforcement_learning/pipeline_reinforcement_learning_vertex_ai/pipeline_reinforcement_learning_vertex_ai.ipynb:614]: Why do we need to copy the dataset? Why not use it form the original location?

**Comment by Ark-kun** [comment]
Got this error when running at the ModelUpload stage: `google.api_core.exceptions.FailedPrecondition: 400 The Cloud Storage bucket of `gs://avolkov/tmp/artifacts` is in location `us`. It must be in the same regional location as the service location `us-central1`.`

/cc @sasha-gitg

**Comment by KathyFeiyang** [comment]
> Got this error when running at the ModelUpload stage: `google.api_core.exceptions.FailedPrecondition: 400 The Cloud Storage bucket of `gs://avolkov/tmp/artifacts`is in location`us`. It must be in the same regional location as the service location `us-central1`.`
> 
> /cc @sasha-gitg

Thanks for pointing this out. I encountered the same issue with [CustomContainerTrainingJob](https://github.com/googleapis/python-aiplatform/blob/v0.8.0/google/cloud/aiplatform/training_jobs.py#L2153), and also logged this in a friction log (which I will send you offline).

For the two notebooks, there are instructions in the GCS bucket creation section about avoiding multi-regional buckets.

File: tutorials/community/reinforcement_learning/pipeline_reinforcement_learning_vertex_ai/src/trainer/trainer_component.py (hunk 2 above)
Comment (KathyFeiyang) [reply tutorials/community/reinforcement_learning/pipeline_reinforcement_learning_vertex_ai/src/trainer/trainer_component.py:31]: In my experimentation, `str` caused an error with reading the training artifact output parameter, whereas `components.OutputPath` worked. I logged this in a friction log and will refer you to the entry.

File: tutorials/community/reinforcement_learning/pipeline_reinforcement_learning_vertex_ai/pipeline_reinforcement_learning_vertex_ai.ipynb (hunk 3 above)
Comment (KathyFeiyang) [reply tutorials/community/reinforcement_learning/pipeline_reinforcement_learning_vertex_ai/pipeline_reinforcement_learning_vertex_ai.ipynb:614]: We don't expect users to download the dataset themselves, although they can. As per [this notebook template](https://github.com/GoogleCloudPlatform/ai-platform-samples/blob/master/ai-platform-unified/notebooks/notebook_template.ipynb) and the instruction "Googlers: Add your dataset to the public samples bucket within gs://cloud-samples-data/ai-platform-unified, if it doesn't already exist there.}", we wanted to host the dataset in a public bucket and add this command to download the dataset for users.

Comment (Ark-kun) [reply tutorials/community/reinforcement_learning/pipeline_reinforcement_learning_vertex_ai/pipeline_reinforcement_learning_vertex_ai.ipynb:614]: It's good that the dataset is public.

What I meant was why should the user copy this public dataset from `gs://cloud-samples-data` to their own bucket?

P.S. The copy can be needed if the BigQuery is incompatible with the `gs://cloud-samples-data` bucket. But in that case it would be great to create/find a public bucket that is compatible.

**Comment by morgandu** [comment]
@KathyFeiyang please add the prototype to the [CODEOWNERS](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/master/docs/CODEOWNERS)

File: tutorials/community/reinforcement_learning/pipeline_reinforcement_learning_vertex_ai/pipeline_reinforcement_learning_vertex_ai.ipynb (hunk 3 above)
Comment (KathyFeiyang) [reply tutorials/community/reinforcement_learning/pipeline_reinforcement_learning_vertex_ai/pipeline_reinforcement_learning_vertex_ai.ipynb:614]: This is a great point. A MovieLens environment from TF-Agents will read the public dataset, not BigQuery. But the environment can indeed load form the public bucket directly. I will make changes to the notebooks.

**Comment by sasha-gitg** [comment]
In Vertex, buckets must be in a single region and must match the region of the Vertex service. We have an open ticket to add verification before we call the API(b/183494969) in the Vertex SDK. In many scenarios the service exception is informative enough (like the example above) and is generally raised immediately. 

For components, I don't think this is feasible because the storage uri could be passed in as PipelineParam and we will not know the identity of the uri until the task is executed.

Are you thinking of a different solution?

File: tutorials/community/reinforcement_learning/README.md (hunk 1 above)
Comment (KathyFeiyang) [reply tutorials/community/reinforcement_learning/README.md]: Thanks for catching this - This is addressed.

File: tutorials/community/reinforcement_learning/pipeline_reinforcement_learning_vertex_ai/src/trainer/trainer_component.py (hunk 2 above)
Comment (KathyFeiyang) [reply tutorials/community/reinforcement_learning/pipeline_reinforcement_learning_vertex_ai/src/trainer/trainer_component.py]: Thanks for pointing this out. I agree that `str` would be the correct choice here.
Deferring the code change to [this ongoing PR](https://github.com/KathyFeiyang/vertex-ai-samples/pull/1/files#r683279722).

**Review by morgandu — APPROVED** [review]
LGTM, thank you @KathyFeiyang for your contribution to our community content!

**Comment by KathyFeiyang** [comment]
@Ark-kun @sasha-gitg Thank you for the discussion above around the potential improvement to using GCS buckets with Vertex. For the scope of this PR and adding the two RL prototypes, the feature change with Vertex is not immediately necessary, because users are instructed in the notebooks to match bucket region with Vertex region, and platform changes can't be implemented in the context of this PR. Therefore, I'll prepare to wrap up the PR. Meanwhile, I do think the discussion is valuable to continue.
